pkg/types/types.go

```diff
@@ -70,12 +116,40 @@ type RetrievalResult struct {
 	Err   error
 }
 
-type CandidateRetrievalCall struct {
-	Candidates         []RetrievalCandidate
-	CandidateRetrieval CandidateRetrieval
+var _ RetrievalTask = AsyncRetrievalTask{}
+
+// AsyncRetrievalTask runs an asynchronous retrieval and returns a result
+type AsyncRetrievalTask struct {
+	Candidates              InboundAsyncCandidates
+	AsyncCandidateRetrieval CandidateRetreival
 }
 
-type RetrievalCoordinator func(context.Context, []CandidateRetrievalCall) (*RetrievalStats, error)
+// Run executes the asychronous retrieval task
+func (acrc AsyncRetrievalTask) Run() (*RetrievalStats, error) {
+	return acrc.AsyncCandidateRetrieval.RetrieveFromAsyncCandidates(acrc.Candidates)
+}
+
+var _ RetrievalTask = DeferredErrorTask{}
+
+// DeferredErrorTask simply reads from an error channel and returns the result as an error
+type DeferredErrorTask struct {
+	Ctx     context.Context
+	ErrChan <-chan error
+}
+
+// Run reads the error channel and returns a result
+func (der DeferredErrorTask) Run() (*RetrievalStats, error) {
+	select {
+	case <-der.Ctx.Done():
+		return nil, der.Ctx.Err()
+	case err := <-der.ErrChan:
```

Comment:
```suggestion
func (det DeferredErrorTask) Run() (*RetrievalStats, error) {
	select {
	case <-det.Ctx.Done():
		return nil, det.Ctx.Err()
	case err := <-det.ErrChan:
```

same thing, not essential this is changed, just adds a tiny speedbump to the cognitive load in grokking the code